sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,502 @@
+    return Float.valueOf(addressedTokens.divide(partitioner.getTotalTokens()).toString());
+  }
+
+  /**
+   * Gets the list of token ranges that a table occupies on a give Cassandra node.
+   *
+   * @param connectionConfiguration The {@link CassandraIO.ConnectionConfiguration} for cluster
+   *     connection
+   * @param table The table whose token ranges are fetched
+   * @return The list of {@link TokenRange}
+   */
+  //compatible Cassandra > 2.1.5
+  private static List<TokenRange> getTokenRanges(
+      Cluster cluster, CassandraIO.ConnectionConfiguration connectionConfiguration, String table) {
+    try (Session session = cluster.newSession()) {
+      ResultSet resultSet =
+          session.execute(
+              "SELECT range_start, range_end, partitions_count, mean_partition_size FROM "
+                  + "system.size_estimates WHERE keyspace_name = ? AND table_name = ?",
+              connectionConfiguration.getKeyspace(),
+              table);
+
+      ArrayList<TokenRange> tokenRanges = new ArrayList<>();
+      for (Row row : resultSet) {
+        TokenRange tokenRange =
+            new TokenRange(
+                row.getLong("partitions_count"),
+                row.getLong("mean_partition_size"),
+                row.getString("range_start"),
+                row.getString("range_end"));
+        tokenRanges.add(tokenRange);
+      }
+      // The table may not contain the estimates yet
+      // or have partitions_count and mean_partition_size fields = 0
+      // if the data was just inserted and the amount of data in the table was small.
+      // This is very common situation during tests,
+      // when we insert a few rows and immediately query them.
+      // However, for tiny data sets the lack of size estimates is not a problem at all,
+      // because we don't want to split tiny data anyways.
+      // Therefore, we're not issuing a warning if the result set was empty
+      // or mean_partition_size and partitions_count = 0.
+      return tokenRanges;
+    } catch (InvalidQueryException e) {
+      LOG.warn(
+          "Failed to fetch size estimates for $keyspaceName.$tableName "
+              + "from system.size_estimates table. "
+              + "The number of created bundles may be inaccurate. "
+              + "Please make sure you use Cassandra 2.1.5 or newer.");
+      ArrayList<TokenRange> tokenRanges = new ArrayList<>();
+      tokenRanges.add(new TokenRange(0L, 0L, "0", "0"));
+      return tokenRanges;
+    }
+  }
+
+  /**
+   * Measure distance between two tokens.
+   *
+   * @param cluster The cassandra {@link Cluster} object
+   * @param token1 The measure is symmetrical so token1 and token2 can be exchanged
+   * @param token2 The measure is symmetrical so token1 and token2 can be exchanged
+   * @return Number of tokens that separate token1 and token2
+   */
+  private static BigInteger distance(Cluster cluster, BigInteger token1, BigInteger token2) {
```

Comment:
We use the Cluster only for the sake of doing new Partitioner(cluster).getTotalTokens(). It'd be cleaner and easier to unit-test if you just pass a "BigInteger totalTokens".